src/runtime_src/core/pcie/linux/shim.cpp

```diff
@@ -131,28 +132,220 @@ namespace xocl {
+    };
+
     uint64_t qhndl;		/* queue handle */
     bool h2c; 			/* queue is for H2C direction */
     bool qAioEn;		/* per queue aio is enabled */
+    bool qAioBatchEn;		/* per queue aio req. batching is enabled */
+    bool qExit;			/* queue is being stopped/released */
     aio_context_t qAioCtx;	/* per queue aio context */
 
+    /* aio batching threshold to flush the queued i/o */
+    unsigned int timer_ms;	/* timer in milisecond */
+    unsigned int byteThresh;	/* total bytes count */
+    unsigned int pktThresh;	/* total buffer/packet count */
+
+    unsigned int byteCnt;	/* total bytes of queued i/o */
+    unsigned int bufCnt;	/* total buffer count of queued i/o */
+    unsigned int cbSubmitCnt;	/* total # submitted i/o */
+    unsigned int cbPollCnt;	/* total # of polled/completed i/o */
+    int cbErrCnt;		/* submission error */
+    int cbErrCode;		/* submission error code */
+
+    std::thread qWorker;	/* aio batching thread */
+    std::mutex reqLock;		/* lock to protect i/o related info. */
+    std::list<struct queued_io> reqList; /* queued up i/o request */
+    std::condition_variable cv; /* for wait/wake up the batching thread */
+
+    /* release queued i/o and update the stats */
+    int release_request(int count)
+    {
+        /* calling function should hold the reqLock */
+        int i = 0;
+        while (i < count && !(reqList.empty())) {
+            auto it = reqList.begin();
+            auto qio = (*it);
+
+	    i++;
+            bufCnt--;
+            byteCnt -= qio.len;
+            reqList.pop_front();
+        }
+        
+        return i;
+    } 
+
+    /* prepare io submission structures */
+    void prepare_io(struct iocb *cb, struct iovec *iov,
+                    struct xocl_qdma_req_header *header,
+		    unsigned long buf_va, unsigned long buf_len,
+                    uint64_t priv_data)
+    {
+        iov[0].iov_base = header;
+        iov[0].iov_len = sizeof(struct xocl_qdma_req_header);
+        iov[1].iov_base = (void *)buf_va;
+        iov[1].iov_len = buf_len;
+
+        if (cb) {
+            memset(cb, 0, sizeof(struct iocb));
+            cb->aio_fildes = (int)qhndl;
+            cb->aio_lio_opcode = h2c ? IOCB_CMD_PWRITEV : IOCB_CMD_PREADV;
+            cb->aio_buf = (uint64_t)iov;
+            cb->aio_offset = 0;
+            cb->aio_nbytes = 2;
+            cb->aio_data = priv_data;
+        }
+    }
+
+    /* submit all of the queued i/o */
+    int queue_flush_aio_request(void)
+    {
+        reqLock.lock();
+
+        /* if there is submission error, wait till all requests are drained */
+        if (cbErrCnt) {
+            reqLock.unlock();
+            return 0;
+        }
+
+        do {
+            /* submit all queued requests */
+            unsigned int cb_max = bufCnt;
+            reqLock.unlock();
+
+            unsigned int cb_cnt = 0;
+            struct iocb **cbpp = (struct iocb **)malloc(cb_max * sizeof(struct iocb *));
```

Comment:
malloc followed by a later free looks unsafe.  Can this be scoped to be exception safe?
Could it maybe be a std::vector?
std::vector<iocb*> cbpp(cb_max);
which is automatically scoped and much preferable over raw new and delete.